addons/binding/org.openhab.binding.exec/src/main/java/org/openhab/binding/exec/handler/ExecHandler.java

```diff
@@ -79,13 +82,17 @@ public void handleCommand(ChannelUID channelUID, Command command) {
                         scheduler.schedule(periodicExecutionRunnable, 0, TimeUnit.SECONDS);
                     }
                 }
-            } else if (channelUID.getId().equals(INPUT)) {
-                if (command instanceof StringType) {
-                    String previousInput = lastInput;
-                    lastInput = command.toString();
-                    if (lastInput != null && !lastInput.equals(previousInput)) {
-                        if (getConfig().get(AUTORUN) != null && ((Boolean) getConfig().get(AUTORUN)).booleanValue()) {
-                            lastInput = command.toString();
+            } else if (channelUID.getId().equals(STRING_INPUT) || channelUID.getId().equals(SWITCH_INPUT)
+                    || channelUID.getId().equals(DIMMER_INPUT) || channelUID.getId().equals(CONTACT_INPUT)
+                    || channelUID.getId().equals(ROLLERSHUTTER_INPUT)) {
+                String previousInput = lastInput;
+                lastInput = command.toString();
+                if (getConfig().get(RUN_ON_INPUT) != null && ((Boolean) getConfig().get(RUN_ON_INPUT)).booleanValue()) {
+                    if (getConfig().get(REPEAT_ENABLED) != null
+                            && ((Boolean) getConfig().get(REPEAT_ENABLED)).booleanValue()) {
+                        scheduler.schedule(periodicExecutionRunnable, 0, TimeUnit.SECONDS);
+                    } else {
+                        if (lastInput != null && !lastInput.equals(previousInput)) {
```

Comment:
The code would be more clear if 'lastInput' was named something like 'lastInputCommand' since it's not clear if the string is the last command (converted to a string) or the last state (converted to a string). However, per my previous comment I think the run channel should use the current item state rather than the last command to the input channel so that would eliminate this variable.